pykafka/balancedconsumer.py

```diff
@@ -428,17 +434,24 @@ def _rebalance(self):
                     # If retrying, be sure to make sure the
                     # partition allocation is correct.
                     participants = self._get_participants()
-                    partitions = self._decide_partitions(participants)
+                    if self._consumer_id not in participants:
+                        # situation that only occurs if our zk session expired
+                        self._add_self()
+                        participants.append(self._consumer_id)
 
-                    old_partitions = self._partitions - partitions
-                    self._remove_partitions(old_partitions)
+                    new_partitions = self._decide_partitions(participants)
 
-                    new_partitions = partitions - self._partitions
-                    self._add_partitions(new_partitions)
+                    # Update zk with any changes:
+                    old_partitions = self._get_held_partitions()
+                    self._remove_partitions(old_partitions - new_partitions)
+                    self._add_partitions(new_partitions - old_partitions)
 
                     # Only re-create internal consumer if something changed.
-                    if old_partitions or new_partitions:
-                        self._setup_internal_consumer()
+                    # Note that `old_partitions` may be different from
+                    # `self._partitions` if our zk session expired (in which
+                    # case the former may be empty while the latter isn't)
```

Comment:
Yeah, I was really struggling with the wording of this. What I mean to highlight is that this line is not the same as
```python
if new_partitions != old_partitions:
```
.... even though 99% of the time it is actually the same - it's only different after we recover from a zk connection loss. Hmm, does that get closer to what I should have written? I'll have another go at it now.